python-client/giskard/scanner/performance.py

```diff
@@ -0,0 +1,235 @@
+        self.params = {}
+        self.model = model
+        self.dataset = dataset
+        self.test_names = test_names
+
+    def run(
+        self,
+        model: Optional[BaseModel] = None,
+        dataset: Optional[Dataset] = None,
+        slicer="tree",
+        test_names: list = None,
+        threshold=-0.1,
+    ):
+        model = model or self.model
+        dataset = dataset or self.dataset
+        test_names = test_names or self.test_names
+        tests = []
+
+        if test_names is None:
+            if model.is_classification:
+                tests.append(test_diff_f1)
+            else:
+                tests.append(test_diff_rmse)
+        else:
+            for test_name in test_names:
+                if test_name == "f1":
+                    tests.append(test_diff_f1)
+                if test_name == "accuracy":
+                    tests.append(test_diff_accuracy)
+                if test_name == "recall":
+                    tests.append(test_diff_recall)
+                if test_name == "precision":
+                    tests.append(test_diff_precision)
+                if test_name == "rmse":
+                    tests.append(test_diff_rmse)
+
+        if model is None:
+            raise ValueError("You need to provide a model to test.")
+        if dataset is None:
+            raise ValueError("You need to provide an evaluation dataset.")
+
+        # …
+
+        # Calculate loss
+        meta = self._calculate_meta(model, dataset)
+        slices = self._find_slices(dataset.select_columns(model.meta.feature_names), model, meta, slicer)
+
+        # Keep only slices of size at least 5% of the dataset
+        slices = [s for s in slices if 0.05 * len(dataset) <= len(dataset.slice(s))]
+
+        issues = self._find_issues(slices, model, dataset, meta, tests, threshold)
+
+        return PerformanceScanResult(issues)
+
+    def _find_issues(self, slices, model, dataset, dataset_meta, tests, threshold):
+        # Use a precooked model to speed up the tests
+        precooked = PrecookedModel.from_model(model, dataset)
+        issues = []
+        for s in slices:
+            for test in tests:
+                test_result = self._diff_test(s, precooked, dataset, test, threshold)
+                if not test_result.passed:
+                    issues.append(Issue(s, model, dataset, dataset_meta, test_result, self._get_test_name(test)))
+
+        return issues
+
+    def _get_test_name(self, test):
+        # @TODO: throw this away when we can
+        if test.meta.name.endswith("f1"):
+            return "F1 score"
+        if test.meta.name.endswith("accuracy"):
+            return "Accuracy"
+        if test.meta.name.endswith("recall"):
+            return "Recall"
+        if test.meta.name.endswith("rmse"):
+            return "RMSE"
+        if test.meta.name.endswith("precision"):
+            return "Precision"
+
+    def _diff_test(self, slice_fn, model, dataset, test_fn, threshold):
```

Comment:
nitpick: this method can be inlined, especially as it's only used once

Reply:
will drop soon